File: BallLog_v1/BallLog_v1/ViewModels/StatsViewModel.swift
import Foundation

final class StatsViewModel: ObservableObject {
    @Published private(set) var total: Int = 0
    @Published private(set) var goals: Int = 0
    @Published private(set) var assists: Int = 0
    @Published private(set) var avgRating: Double = 0
    
    private let repository: MatchRepository
    
    init(repository: MatchRepository) {
        self.repository = repository
        reload()
    }
    
    func reload() {
        let list = repository.list()
        total = list.count
        goals = list.map(\.goals).reduce(0, +)
        assists = list.map(\.assists).reduce(0, +)
        if total > 0 {
            let sum = list.map(\.rating).reduce(0, +)
            avgRating = (Double(sum) / Double(total) * 10).rounded() / 10
        } else {
            avgRating = 0
        }
    }
}

File: src/mvvn/viewmodel/statsViewModel.ts
import { useEffect, useMemo, useState } from "react";
import { matchRepository } from "@mvvn/data/matchRepository";
import { Match } from "@mvvn/model/match";

export function useStatsViewModel() {
  const [items, setItems] = useState<Match[]>([]);

  useEffect(() => {
    (async () => {
      const list = await matchRepository.list();
      setItems(list);
    })();
  }, []);

  const stats = useMemo(() => {
    const total = items.length;
    const goals = items.reduce((sum, match) => sum + match.goals, 0);
    const assists = items.reduce((sum, match) => sum + match.assists, 0);
    const avgRating = total ? Math.round((items.reduce((sum, match) => sum + match.rating, 0) / total) * 10) / 10 : 0;
    const avgDuration = total ? Math.round(items.reduce((sum, match) => sum + match.durationMin, 0) / total) : 0;
    const productivity = avgDuration ? Math.round((goals / total) * 100) / 10 : 0;
    return { total, goals, assists, avgRating, avgDuration, productivity };
  }, [items]);

  const timeline = useMemo(() => items.slice().reverse(), [items]);

  return { stats, timeline };
}
